lib/src/rules/avoid_shadowing.dart

```diff
@@ -0,0 +1,204 @@
+// Copyright (c) 2018, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library linter.src.rules.avoid_shadowing;
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:linter/src/analyzer.dart';
+
+const desc = r'Do not use a name already visible.';
+
+const details = r'''
+
+**DO** Do not use a name already visible as getter or variable.
+
+**BAD:**
+```
+var k = null;
+
+class A {
+  var a;
+}
+
+class B extends A {
+  get b => null;
+
+  m() {
+    var a; // LINT
+    var b; // LINT
+    var k; // LINT
+  }
+}
+```
+
+**GOOD:**
+```
+var k = null;
+
+class A {
+  var a;
+}
+
+class B extends A {
+  get b => null;
+
+  m() {
+    var c; // OK
+    var d; // OK
+    var e; // OK
+  }
+}
+```
+''';
+
+class AvoidShadowing extends LintRule implements NodeLintRule {
+  AvoidShadowing()
+      : super(
+            name: 'avoid_shadowing',
+            description: desc,
+            details: details,
+            group: Group.errors);
+
+  @override
+  void registerNodeProcessors(NodeLintRegistry registry) {
+    final visitor = new _Visitor(this);
+    registry.addVariableDeclarationStatement(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+
+  _Visitor(this.rule);
+
+  @override
+  visitVariableDeclarationStatement(VariableDeclarationStatement node) {
+    final variables = node.variables.variables.toList();
+    final library = variables.first.declaredElement.library;
+
+    // exclude pattern : var name = this.name;
+    variables.removeWhere((variable) {
+      final initializer = variable.initializer;
+      return initializer is PropertyAccess &&
+          initializer.propertyName.name == variable.name.name &&
+          (initializer.target is ThisExpression ||
+              initializer.target is SuperExpression);
+    });
+
+    // exclude pattern : same name as current getter name
+    var currentGetter = _getCurrentGetter(node);
+    if (currentGetter != null) {
+      variables.removeWhere(
+          (variable) => currentGetter.name.name == variable.name.name);
+    }
+
+    bool skipInstanceMembers = false;
+    AstNode current = node;
+    while (current != null) {
```

Comment:
This doesn't seem very efficient, but perhaps in practice it doesn't matter. Have we done any timing measures for this rule?

One easy improvement would be to return before line 98 if `variables` is empty.